### ISSUE TITLE
Feature Request: Select links with ivy rather than avy

### ISSUE THREAD

**Issue by rgrinberg** [noctuid/link-hint.el#38]
Sometimes, it's more convenient to find the link/button by entering a substring rather than entering an avy code. It also makes it possible to select elements that are off the screen this way. Would it be possible something to have something like akin to `counsel-link-hint`?

**Comment by noctuid** [comment]
Yes, I've mentioned wanting to add support for this in a comment in #14. It should be pretty easy. Run `link-hint--collect` to get candidates and then pass them to `completing-read` (so it would work with other completion frameworks) and then call the action on them. I'll see if it's something I can do quickly next week.

**Comment by noctuid** [comment]
I had some time to work on this yesterday evening and also implemented #14. See the `completing-read` branch if you want to try it and give feedback. It is actually instant with very large buffers for me, but I don't have files with a ton of links (didn't test anything with more than 150 links).

Note to self on what is left to do before merging:
- [ ] Test everything to make sure nothing broke
- [ ] Maybe read list of strings instead of symbols for actions
- [ ] Test all-windows-alt; use with completing-read? remove corresponding todos from link-hint.el
- [ ] Ensure every entry to completing-read is unique (add window and position in some visually pleasing... or invisible way); remove corresponding todo

**Comment by rgrinberg** [comment]
Looking good. How would one write an ivy action that would navigate to the point where the link is?

Also, do you think it makes sense to add better integration with ivy if ivy is loaded? While it's nice to make things generic, the majority of emacs will not customize things and I feel like it's better to make things as nice as possible out of the box.

**Comment by noctuid** [comment]
Right now, you could set `link-hint-restore` to nil (for the duration of a command) and use `:copy` to copy and jump to a link. I'll have to add a dedicated action or way to just jump.

> Also, do you think it makes sense to add better integration with ivy if ivy is loaded? 

Is there some ivy-specific feature you want? If `ivy-mode` is on, it will use ivy.

**Comment by rgrinberg** [comment]
>  I'll have to add a dedicated action or way to just jump.

A dedicated action would be nice.

> Is there some ivy-specific feature you want? If ivy-mode is on, it will use ivy.

I was thinking of adding some fancier previewing using ivy-rich. I'll experiment a little more and report back.

**Comment by minad** [comment]
@noctuid Do you plan to add the `completing-read` interface to link-hint? It is quite useful also in combination with `embark-collect`.

**Comment by noctuid** [comment]
After finding `ffap-menu`, I wasn't how sure how useful this is since that handles a lot of links, and in some buffers you can get a lot of garbage with the link hint version (not all links/buttons have useful/searchable text). That is a fixable issue though. Most of the work is already done, so if someone finds it useful, I'd be happy to update and merge the branch. I'd probably like someone to give more feedback on it though because I don't think I will often use it myself.

**Comment by minad** [comment]
Interesting, I was not aware of `ffap-menu`. But a link-hint `completing-read` interface should be more general? cc @oantolin

**Comment by nicolas-graves** [comment]
After finding about `ffap-menu`, I was wondering if functionalities of `link-hint`can actually be reproduced with `completing-read` and `ffap` regex url handling functionalities alone (without `avy`-like functionality then, and with an extensive `ffap-alist`). Do you see something hindering it ? I might consider writing such a `ffap-alist` configuration. @minad @oantolin @noctuid

**Comment by noctuid** [comment]
Not sure if I understand exactly what functionality you want, but a large number of link-hint links/buttons are found using text or overlay properties, not regexp, so you couldn't replicate link-hint with an alternate completing-read interface with regexp alone.

**Comment by nicolas-graves** [comment]
Sorry, I wasn't clear, I've read again the help for the variable `ffap-alist`. It works with proper functions using either current major modes or regexp. I'm not familiar enough with text or overlay properties, but if they fit in a function used when in a major-mode, it might be reproducible. I will take a look at it next month.
